src/libraries/System.IO.Compression/src/System/IO/Compression/ZipCryptoStream.cs

```diff
@@ -0,0 +1,391 @@
+        {
+            ArgumentNullException.ThrowIfNull(baseStream);
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var stream = new ZipCryptoStream(baseStream, encrypting: true, leaveOpen, passwordVerifierLow2Bytes, crc32);
+            stream.InitKeysFromBytes(password.Span);
+
+            // No async work needed for encryption constructor
+            return Task.FromResult(stream);
+        }
+
+        private void EnsureHeader()
+        {
+            if (!_encrypting || _headerWritten) return;
+
+            Span<byte> hdrPlain = stackalloc byte[12];
+
+            // bytes 0..9 are random
+            RandomNumberGenerator.Fill(hdrPlain.Slice(0, 10));
+
+            // bytes 10..11: check bytes (CRC-based if crc32 provided; else DOS time low word)
+            if (_crc32ForHeader.HasValue)
+            {
+                uint crc = _crc32ForHeader.Value;
+                hdrPlain[10] = (byte)((crc >> 16) & 0xFF);
+                hdrPlain[11] = (byte)((crc >> 24) & 0xFF);
+            }
+            else
+            {
+                hdrPlain[10] = (byte)(_verifierLow2Bytes & 0xFF);
+                hdrPlain[11] = (byte)((_verifierLow2Bytes >> 8) & 0xFF);
+            }
+
+            // Encrypt & write; update keys with PLAINTEXT per spec
+            byte[] hdrCiph = new byte[12];
+            for (int i = 0; i < 12; i++)
+            {
+                byte ks = DecipherByte();
+                byte p = hdrPlain[i];
+                hdrCiph[i] = (byte)(p ^ ks);
+                UpdateKeys(p);
+            }
+
+            _base.Write(hdrCiph, 0, 12);
+            _headerWritten = true;
+            _position += 12;
+        }
+
+        private async ValueTask EnsureHeaderAsync(CancellationToken cancellationToken)
+        {
+            if (!_encrypting || _headerWritten) return;
+
+            byte[] hdrPlain = new byte[12];
+
+            // bytes 0..9 are random
+            RandomNumberGenerator.Fill(hdrPlain.AsSpan(0, 10));
+
+            // bytes 10..11: check bytes (CRC-based if crc32 provided; else DOS time low word)
+            if (_crc32ForHeader.HasValue)
+            {
+                uint crc = _crc32ForHeader.Value;
+                hdrPlain[10] = (byte)((crc >> 16) & 0xFF);
+                hdrPlain[11] = (byte)((crc >> 24) & 0xFF);
+            }
+            else
+            {
+                hdrPlain[10] = (byte)(_verifierLow2Bytes & 0xFF);
+                hdrPlain[11] = (byte)((_verifierLow2Bytes >> 8) & 0xFF);
+            }
+
+            // Encrypt & write; update keys with PLAINTEXT per spec
+            byte[] hdrCiph = new byte[12];
+            for (int i = 0; i < 12; i++)
+            {
+                byte ks = DecipherByte();
+                byte p = hdrPlain[i];
+                hdrCiph[i] = (byte)(p ^ ks);
+                UpdateKeys(p);
+            }
+
+            await _base.WriteAsync(hdrCiph.AsMemory(0, 12), cancellationToken).ConfigureAwait(false);
```

Comment:
Most of the code in this function deals with the header construction, you can eliminate the duplication by having a function like CalculateHeader(Span<byte>) and do the actual writes from Write/WriteAsync